### PR TITLE
[voice] Support managed dialogs

### PR DESCRIPTION
**Pull request by GiviMAD**
This PR adds a new functionality to the voice manager for register/unregister dialog descriptions.
The voice manager will try to keep a started dialog for each registered description (and stop the dialog if it's unregistered).
The dialog descriptions are persisted using a storage.

This PR do not introduce any code breaking changes, but introduces the behavior to stop the dialogs whenever a service in use became unavailable. 

Related to https://github.com/openhab/openhab-core/issues/3265

<s>This is a draft because integration tests are missing.</s>

Signed-off-by: Miguel Álvarez <miguelwork92@gmail.com>

### CODE REVIEW

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java [hunk 1]
```diff
@@ -951,8 +1038,92 @@ private Comparator<Voice> createVoiceComparator(Locale locale) {
...
+            if (dialogsToStop.isEmpty()) {
+                return;
+            }
+            for (var dialog : dialogsToStop) {
+                stopDialog(dialog.dialogContext.source());
+            }
+        }
+    }
+
+    private void debounceBuildDialogRegistrations() {
```
Comment (J-N-K) [inline bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java:1053]: Why is this needed?

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/DialogRegistration.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
```
Comment (J-N-K) [inline bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/DialogRegistration.java:2]:
```suggestion
 * Copyright (c) 2010-2023 Contributors to the openHAB project
```

**Review by J-N-K — CHANGES_REQUESTED** [review]
What is the use-case for that?

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java (hunk 1 above)
Comment (GiviMAD) [reply bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java:1053]: It's not really needed, but I think it will skip a lot of unnecessary executions, as this needs to be called on every related add-on registration but there is not need to do it immediately, I added it thinking in the following cases:

* On the initial start of openHAB this method will be called a lot of times.
* Normally the sink and source related to the dialog will bellow to the same add-on so registration will happen mostly at same time.

Let me know if you think is better to remove/improve it.

**Comment by GiviMAD** [comment]
> What is the use-case for that?

For the new methods? The final use-case is exposing a way users can configure dialogs that openHAB will keep running whenever is possible event after an openHAB reboot.

I'll explain in more detail what is my point of view after my experience and some feedback on the marketplace:

As an add-on developer the current methods startDialog and stopDialog are perfect to consume , I'm using those on the HABSpeaker add-on you can find at the marketplace, where I'm playing with audio recognition on the browser and testing other ideas.

But as an openHAB user or an UI client the startDialog and stopDialog functionality is not really useful, because what those consumer want is to be able to configure the desired services for a dialog and that the system keep it running whenever is possible (this involves the initial dialog start, and restart the dialog after a reboot or a related bundle update), be able to check its state, and unregister it whenever you don't what it any more.

So my proposal for solving this is adding these new methods that will be exposed through the console but also to the API hopping some UI contributor is interested in build a configuration page for the dialogs which I think can consist on a list of dialog registrations and a creation form that can be pretty similar to the voice options configuration page. This way the user will be able to manage its dialogs from the UI.

And thank you for the review! :)

**Comment by GiviMAD** [comment]
Of course let me know if you think there is a better way for adding this functionality. If this looks ok to you I can add a more tests because right now I've just included a basic one that do not tests all the scenarios.

**Comment by GiviMAD** [comment]
@J-N-K did you have a change to look at this?

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java [hunk 3]
```diff
@@ -951,8 +1039,92 @@ private Comparator<Voice> createVoiceComparator(Locale locale) {
...
+    }
+
+    private void tryBuildDialogRegistration(DialogRegistration dialogRegistration) {
+        var builder = getDialogContextBuilder().withSink(audioManager.getSink(dialogRegistration.sinkId))
+                .withSource(audioManager.getSource(dialogRegistration.sourceId));
+        String ksId = dialogRegistration.ksId;
+        if (ksId != null) {
+            builder.withKS(getKS(ksId));
+        }
+        String keyword = dialogRegistration.keyword;
```
Comment (J-N-K) [inline bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java:1079]: You do a lot of these null-checks in several places. IMO it would be better to accept `null` in the builder and discard the value there. You can also improve `getKS` and similar:

```
    @Override
    public @Nullable KSService getKS(@Nullable String id) {
        return id == null ? null : ksServices.get(id);
    }
```

then you can really make good use of the builder and concatenate the `.withXY` calls.

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java (hunk 1 above)
Comment (J-N-K) [reply bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java:1054]: Understood. Looks useful. Maybe it's better to call it `scheduleDialogRegistrations` and add JavaDoc explaining that each call re-schedules the build.

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java [hunk 4]
```diff
@@ -700,6 +714,50 @@ public void listenAndAnswer(DialogContext context) throws IllegalStateException
...
+    public void unregisterDialog(DialogRegistration registration) {
+        unregisterDialog(registration.sourceId);
+    }
+
+    @Override
+    public void unregisterDialog(String sourceId) {
+        synchronized (dialogRegistrationStorage) {
+            var registrationRef = dialogRegistrationStorage.remove(sourceId);
+            if (registrationRef != null) {
+                dialogRegistrationStorage.remove(sourceId);
```
Comment (J-N-K) [inline bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java:739]: Why do you remove it twice?

**Review by J-N-K — CHANGES_REQUESTED** [review]
Sorry for responding so late. I completely forgot about this. I have left some comments, otherwise LGTM.

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java (hunk 4 above)
Comment (GiviMAD) [reply bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java]: Not needed, my fault. Thank you!

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java (hunk 3 above)
Comment (GiviMAD) [reply bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java:1091]: I don't like it either, but I didn't know how to solved.
I think I have understand your solution looks great, I'll go for it.

Comment (GiviMAD) [reply bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java]: Looks much better now, thank you for proposing this change.
I can take advantage of it also on this one https://github.com/openhab/openhab-core/pull/3253.

**Comment by GiviMAD** [comment]
Sorry, I saw an error and I've added a commit after asking for the review.

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/DialogContext.java [hunk 5]
```diff
@@ -137,17 +137,23 @@ public Builder withSink(@Nullable AudioSink sink) {
         }
 
         public Builder withKS(@Nullable KSService service) {
-            this.ks = service;
+            if (service != null) {
+                this.ks = service;
+            }
             return this;
```
Comment (J-N-K) [inline bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/DialogContext.java:143]: I wonder if you need to do a check at all here. The field is annotated as `@Nullable` anyway, so setting it to `null` will not change anything. Or is there a place where a valid value could be overwritten?

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java [hunk 6]
```diff
@@ -951,8 +1040,73 @@ private Comparator<Voice> createVoiceComparator(Locale locale) {
...
+     * this method schedules a call to {@link #buildDialogRegistrations() buildDialogRegistrations} in five seconds
+     * and cancel the previous scheduled call if any.
+     */
+    private void scheduleDialogRegistrations() {
+        if (dialogRegistrationFuture != null) {
+            dialogRegistrationFuture.cancel(false);
+        }
+        dialogRegistrationFuture = scheduledExecutorService.schedule(this::buildDialogRegistrations, 5,
+                TimeUnit.SECONDS);
+    }
```
Comment (J-N-K) [inline bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java:1066]:
```suggestion
    private void scheduleDialogRegistrations() {
        ScheduledFuture<?> job = this.dialogRegistrationFuture;
        if (job != null) {
            job.cancel(false);
        }
        dialogRegistrationFuture = scheduledExecutorService.schedule(this::buildDialogRegistrations, 5,
                TimeUnit.SECONDS);
    }
```

**Review by J-N-K — COMMENTED** [review]
Otherwise LGTM

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/DialogContext.java (hunk 5 above)
Comment (GiviMAD) [reply bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/DialogContext.java:143]: Yes, that is the reason, it contains the default "keyword spotting" service if any, the reason for it to be nullable is that the listenAndAnswer functionality can be used without it.

File: bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java [hunk 7]
```diff
@@ -151,7 +164,13 @@ protected void activate(BundleContext bundleContext, Map<String, Object> config)
 
     @Deactivate
     protected void deactivate() {
-        stopAllDialogs();
+        dialogProcessors.values().forEach(DialogProcessor::stop);
+        dialogProcessors.clear();
+        var dialogRegistrationFuture = this.dialogRegistrationFuture;
```
Comment (J-N-K) [inline bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java:169]: I'm really sorry, I didn't notice before and have another request. We avoid using `var`. Since this is only cosmetic, it should be a fast-fix. I'll merge afterwards, promised.
